### ISSUE TITLE
False positive metaclass conflict error

### ISSUE THREAD

**Issue by KoltesDigital** [microsoft/pyright#9430]
**Describe the bug**

My goal is to have a list of different things in a JSON file. I rely on Pydantic to parse the content. I don't want to use `isinstance`, rather all these different things implement a common protocol to make their own stuff happen.

**Code or Screenshots**

The following code was passing with version **1.1.371**, but fails with version **1.1.388** (I haven't tested inbetween) directly in command-line (`python -m pyright ...`).

```python
from typing import List, Union

from pydantic import BaseModel, Field, TypeAdapter
from typing_extensions import Annotated, Literal, Protocol, override


class Animal(Protocol):
    def greets(self) -> str: ...


class _CatMeta(type(BaseModel), type(Animal)):
    pass


class Cat(
    BaseModel,
    Animal,
    metaclass=_CatMeta,
):
    kind: Literal["cat"] = "cat"

    @override
    def greets(self) -> str:
        return "meow"


class _DogMeta(type(BaseModel), type(Animal)):
    pass


class Dog(
    BaseModel,
    Animal,
    metaclass=_DogMeta,
):
    kind: Literal["dog"] = "dog"

    @override
    def greets(self) -> str:
        return "woof"


Animals = TypeAdapter(List[Annotated[Union[Cat, Dog], Field(discriminator="kind")]])


def test_can_greet() -> None:
    animals = Animals.validate_python(
        [
            {
                "kind": "cat",
            },
            {
                "kind": "dog",
            },
        ],
    )

    greetings = [animal.greets() for animal in animals]
    assert greetings == [
        "meow",
        "woof",
    ]
```

Errors:

```
  test.py:15:7 - error: The metaclass of a derived class must be a subclass of the metaclasses of all its base classes
    Metaclass "_CatMeta" conflicts with "ModelMetaclass" (reportGeneralTypeIssues)
  test:31:7 - error: The metaclass of a derived class must be a subclass of the metaclasses of all its base classes
    Metaclass "_DogMeta" conflicts with "ModelMetaclass" (reportGeneralTypeIssues)
```

**Comment by erictraut** [comment]
Here's a simplified example, for my own benefit.

```python
class Meta(type): ...

class Base(metaclass=Meta): ...

class A(Base, metaclass=type(Base)): ...
```

What you're doing here is pretty atypical. I'll need to investigate further to see if this is something that pyright can support. For comparison, mypy doesn't support dynamic base types at all.

**Comment by KoltesDigital** [comment]
@erictraut Indeed, but I thought maybe it was specific to Pedantic, because BaseModel has an explicit metaclass.

Besides, I'm curious: how would you do, if my design is atypical? I agree that it's not true polymorphism as the precise list of types has to be known, but I'd prefer avoid isinstance chains, because afaik I can't ensure that all variants have been tested. I see my design as a light visitor pattern, with no composite.

**Comment by erictraut** [comment]
It's atypical because you're using dynamic expressions (`type(BaseModel )` and `type(Animal)`) as base classes rather than specifying classes by name (e.g. `ModelMetaclass`). As I mentioned, other type checkers like mypy don't support this at all.

If you change my minimal repro to specify the metaclass directly rather than using a dynamic expression, the error goes away:

```python
class Meta(type): ...
class Base(metaclass=Meta): ...
class A(Base, metaclass=Meta): ...
```

Similarly, with your example, this eliminates the error in pyright:

```python
from pydantic import BaseModel
from pydantic._internal._model_construction import ModelMetaclass
from typing_extensions import Protocol

class Animal(Protocol): ...
class _CatMeta(ModelMetaclass, type(Animal)): ...
class Cat(BaseModel, Animal, metaclass=_CatMeta): ...
```

I'll point out also that you're relying on details of two classes (pydantic's `ModelMetaclass` and the metaclass for `Protocol`) that are not documented and should therefore be considered internal implementation details. `ModelMetaclass` is defined in an `_internal` submodule and is (presumably intentionally) not exposed as part of pydantic's supported interface contract. By deriving from it directly, you may be setting yourself up for incompatibilities in the future if its implementation changes.

I don't know enough about what you're doing to offer advice on alternative approaches. There are many dynamic things you can do with Python that will defeat static type checking. That doesn't mean these dynamic things are necessarily wrong, but it means that you may need to forego the benefits of static type checking if you use such techniques.

**Comment by beauxq** [comment]
To me this looks like the problem is mixing structural and nominal subtyping.
My suggestion would be to choose 1, structural or nominal.

If you want structural:
```python
class Animal(Protocol):
    def greets(self) -> str:
        ...


class Cat(
    BaseModel,
):
    kind: Literal["cat"] = "cat"

    def greets(self) -> str:
        return "meow"


class Dog(
    BaseModel,
):
    kind: Literal["dog"] = "dog"

    def greets(self) -> str:
        return "woof"
```

If you want nominal:
```python
class Animal(abc.ABC):
    @abc.abstractmethod
    def greets(self) -> str:
        ...


class Cat(
    BaseModel,
    Animal,
):
    kind: Literal["cat"] = "cat"

    @override
    def greets(self) -> str:
        return "meow"


class Dog(
    BaseModel,
    Animal,
):
    kind: Literal["dog"] = "dog"

    @override
    def greets(self) -> str:
        return "woof"
```

**Comment by KoltesDigital** [comment]
@erictraut well I wasn't really aware there was a metaclass to BaseModel before opening this ticket. Thanks for the discussion, now I understand better what's going on. And I agree it's bad to rely on private details. Ultimately, it seems it's just impossible to inherit from BaseModel and another class that is not ABC.

@beauxq thanks for the suggestions. Don't take it personally, I'm fan of neither, surely because I'm more into system languages. With the first one, I feel an info known to the programmer is not explicited into the code. For instance, if I were to add a method to the protocol, I expect there is a tool to complain right now about implementors not being compliant. Admittedly, they should end up pin some unit tests and issues would be raised at these locations, but that's not so immediate. With the second one, well I want to use Protocol because I think it's cleaner than ABC, so it feels like regression.

In the meantime, I found out there's an alternative way to define Pydantic-ish classes with its own drop-in replacement for `dataclass`, which then allows for any inheritance, and they can be aggregated into a BaseModel / TypeAdapter if needed. For the record, here how I'm going to change my code, which still passes the test and pleases Pyright.

```python
...
from pydantic.dataclasses import dataclass


class Animal(Protocol):
    def greets(self) -> str: ...


@dataclass
class Cat(Animal):
    kind: Literal["cat"] = "cat"

    @override
    def greets(self) -> str:
        return "meow"


@dataclass
class Dog(Animal):
    kind: Literal["dog"] = "dog"

    @override
    def greets(self) -> str:
        return "woof"
```

Anyway, that's very specific to Pydantic. I'm closing the ticket.

**Comment by erictraut** [comment]
I'm going to reopen the issue because it turns out that your code exposes a legitimate bug in pyright. It wasn't properly handling the case where an expression of form `type(<some class>)` is used in a base class list or as a metaclass. With this bug fixed, the first code example in this thread now type checks without error. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.389
